File: trip-kit-booking/src/test/java/com/skedgo/android/tripkit/booking/viewmodel/DateTimeFieldViewModelImplTest.java
package com.skedgo.android.tripkit.booking.viewmodel;

import android.os.Parcel;

import com.skedgo.android.tripkit.booking.BuildConfig;
import com.skedgo.android.tripkit.booking.DateTimeFormField;
import com.skedgo.android.tripkit.booking.TestRunner;

import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.annotation.Config;

import java.util.Calendar;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(TestRunner.class)
@Config(constants = BuildConfig.class)
public class DateTimeFieldViewModelImplTest {
  /**
   * FIXME: Temporarily ignore as it's flaky by the use of {@link Calendar#getInstance()}.
   */
  @Ignore
  @Test
  public void shouldReflectSomeValuesFromField() {
    final DateTimeFormField field = new DateTimeFormField();
    field.setValue(1234567890);
    field.setTitle("Time");
    final DateTimeFieldViewModelImpl viewModel = DateTimeFieldViewModelImpl.create(field);

    assertThat(viewModel.getTitle())
        .describedAs("Should reflect title from field")
        .isEqualTo("Time");
    assertThat(viewModel.getDate())
        .describedAs("Should format date from field")
        .isEqualTo("Sat, 14 Feb 2009");
    assertThat(viewModel.getTime())
        .describedAs("Should format time from field")
        .isEqualTo("06:31");

    assertThat(viewModel.getDay()).isEqualTo(14);
    assertThat(viewModel.getMonth()).isEqualTo(1);
    assertThat(viewModel.getYear()).isEqualTo(2009);
    assertThat(viewModel.getHour()).isEqualTo(6);
    assertThat(viewModel.getMinute()).isEqualTo(31);
  }

  /**
   * FIXME: Temporarily ignore as it's flaky by the use of {@link Calendar#getInstance()}.
   */
  @Ignore
  @Test
  public void shouldDisplayHourIn24hFormat() {
    final DateTimeFormField field = new DateTimeFormField();
    field.setValue(1445235929);
    field.setTitle("Time");

    final DateTimeFieldViewModelImpl viewModel = DateTimeFieldViewModelImpl.create(field);
    assertThat(viewModel.getHour()).isEqualTo(13);
  }

  @Test
  public void shouldBeParcelledProperly() {
    final DateTimeFormField field = new DateTimeFormField();
    field.setValue(1234567890);
    field.setTitle("Time");
    final DateTimeFieldViewModelImpl expected = DateTimeFieldViewModelImpl.create(field);

    final Parcel parcel = Parcel.obtain();
    expected.writeToParcel(parcel, 0);
    parcel.setDataPosition(0);

    final DateTimeFieldViewModelImpl actual = DateTimeFieldViewModelImpl.CREATOR.createFromParcel(parcel);
    assertThat(actual.getTitle())
        .describedAs("Title should be parcelled properly")
        .isEqualTo(expected.getTitle());
    assertThat(actual.getDate())
        .describedAs("Date should parcelled properly")
        .isEqualTo(expected.getDate());
    assertThat(actual.getTime())
        .describedAs("Time should parcelled properly")
        .isEqualTo(expected.getTime());
  }

  @Test
  public void shouldShowSelectedValueProperly() {
    final DateTimeFormField field = new DateTimeFormField();
    field.setValue(1234567890);
    field.setTitle("Time");

    final DateTimeFieldViewModelImpl viewModel = DateTimeFieldViewModelImpl.create(field);
    viewModel.setTime(23, 59);
    viewModel.setDate(2015, 10, 20);

    assertThat(viewModel.getDate())
        .describedAs("Should reflect new date selected")
        .isEqualTo("Fri, 20 Nov 2015");
    assertThat(viewModel.getTime())
        .describedAs("Should reflect new time selected")
        .isEqualTo("23:59");
  }
}

File: trip-kit-booking/src/test/java/com/skedgo/tripkit/booking/viewmodel/DateTimeFieldViewModelImplTest.kt
package com.skedgo.tripkit.booking.viewmodel

import android.os.Parcel
import com.skedgo.tripkit.booking.DateTimeFormField
import com.skedgo.tripkit.booking.viewmodel.DateTimeFieldViewModelImpl.CREATOR.createFromParcel
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.runs
import junit.framework.TestCase.assertEquals
import org.assertj.core.api.Java6Assertions
import org.junit.Ignore
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.util.Calendar

@RunWith(RobolectricTestRunner::class)
class DateTimeFieldViewModelImplTest {

    /**
     * FIXME: Temporarily ignore as it's flaky by the use of {@link Calendar#getInstance()}.
     */
    @Ignore
    @Test
    fun shouldReflectSomeValuesFromField() {
        val field = DateTimeFormField().apply {
            setValue(1234567890)
            title = "Time"
        }
        val viewModel = DateTimeFieldViewModelImpl.create(field)

        assertEquals("Time", viewModel.getTitle(), "Should reflect title from field")
        assertEquals("Sat, 14 Feb 2009", viewModel.getDate(), "Should format date from field")
        assertEquals("06:31", viewModel.getTime(), "Should format time from field")
        assertEquals(14, viewModel.getDay())
        assertEquals(1, viewModel.getMonth())
        assertEquals(2009, viewModel.getYear())
        assertEquals(6, viewModel.getHour())
        assertEquals(31, viewModel.getMinute())
    }

    /**
     * FIXME: Temporarily ignore as it's flaky by the use of {@link Calendar#getInstance()}.
     */
    @Ignore
    @Test
    fun shouldDisplayHourIn24hFormat() {
        val field = DateTimeFormField().apply {
            setValue(1445235929)
            title = "Time"
        }
        val viewModel = DateTimeFieldViewModelImpl.create(field)
        assertEquals(13, viewModel.getHour())
    }

    @Test
    fun shouldBeParcelledProperly() {
        val field = DateTimeFormField().apply {
            setValue(1234567890)
            title = "Time"
        }
        val expected = DateTimeFieldViewModelImpl.create(field)

        // Mock the Parcel
        val parcel = mockk<Parcel>(relaxed = true)

        // Mock the behavior for writing Parcelable
        every { parcel.writeParcelable(any(), any()) } just Runs
        every { parcel.writeSerializable(any()) } just Runs

        // Perform writeToParcel with mocked Parcel
        expected.writeToParcel(parcel, 0)

        // Mock reading from Parcel - this step is crucial when you test the CREATOR
        every { parcel.readParcelable<DateTimeFormField>(any<ClassLoader>()) } returns field
        every { parcel.readSerializable() } returns Calendar.getInstance()

        // Set the data position to zero (simulates resetting the Parcel for reading)
        every { parcel.setDataPosition(0) } just Runs

        // Use the CREATOR to create a new instance from the Parcel
        val actual = DateTimeFieldViewModelImpl.CREATOR.createFromParcel(parcel)

        // Perform assertions
        assertEquals("Title should be parcelled properly", expected.getTitle(), actual.getTitle())
        assertEquals("Date should parcelled properly", expected.getDate(), actual.getDate())
        assertEquals("Time should parcelled properly", expected.getTime(), actual.getTime())
    }

    @Test
    fun shouldShowSelectedValueProperly() {
        val field = DateTimeFormField().apply {
            setValue(1234567890)
            title = "Time"
        }
        val viewModel = DateTimeFieldViewModelImpl.create(field)
        viewModel.setTime(23, 59)
        viewModel.setDate(2015, 10, 20)

        assertEquals("Should reflect new date selected","Fri, 20 Nov 2015", viewModel.getDate())
        assertEquals("Should reflect new time selected", "23:59", viewModel.getTime())
    }
}